### ISSUE TITLE
undefined method '[]' for nil:NilClass

### ISSUE THREAD

**Issue by confact** [gottfrois/grape-attack#5]
Hi @gottfrois!
I am trying to implement grape-attack to my api now.

i get this error when i add `use Grape::Attack::Throttle` to Apiv2 class

``` ruby
F, [2015-11-25T20:19:56.890773 #11763] FATAL -- :
NoMethodError (undefined method `[]' for nil:NilClass):
  app/api/apiv2.rb:22:in `apitoken'
  app/api/v2/stations/stations.rb:30:in `block (2 levels) in <class:Stations>'
```

apiv2.rb:22:

``` ruby
      def apitoken
        Apikey.eager_load(:User).where(apikey: params[:token]).first
      end
```

As soon as I commenting `use Grape::Attack::Throttle` it works again.
What can be the issue?

**Comment by gottfrois** [comment]
`params` seems to be `nil`, not sure it's related to grape-attack

**Comment by confact** [comment]
@gottfrois the params are set.
the url is `http://localhost:3000/api/v2/stations?token=afasfsfafa`
and params is shown in the error as 

```
Request

Parameters:

{"token"=>"afasfsfafa"}
```

and again, when i set # before `use Grape::Attack::Throttle` (commenting it), it works again. So it is either grape-attack or a something getting a conflict here.

**Comment by gottfrois** [comment]
right, can you print the `params` variable just before this line?

```
Apikey.eager_load(:User).where(apikey: params[:token]).first
```

I want to make sure that params is nil so that i can look for such conflict in the gem

**Comment by confact** [comment]
@gottfrois okay, seems to be params being nil, even if it shows in the error page. But when i turn off `use Grape::Attack::Throttle` by commenting it, it works and params are not nil. So i think grape-attack do something to the params. Like middleware or something.

**Comment by confact** [comment]
When commenting out Grape::Attack::Throttle
i get this:

``` ruby
--- !ruby/hash:Hashie::Mash
token: afasfsfafa
page: 1
per_page: 20
offset: 0
```

when i remove the comment and run Grape::Attack::Throttle i get nil.

**Comment by gottfrois** [comment]
alrigth, i'll have a closer look

**Comment by confact** [comment]
@gottfrois great. I will gladly help you with whatever i can do. I can give you my Gemfile.lock so you maybe can see if it is any conflict or such with other gem or new version of a gem breaking something up.

**Comment by confact** [comment]
Here is at least a full backtrace:

```
/vagrant/app/api/apiv2.rb:24:in `apitoken'
/vagrant/app/api/v2/stations/stations.rb:10:in `block (2 levels) in <class:Stations>'
/home/vagrant/.rvm/gems/ruby-2.1.4/bundler/gems/grape-attack-a50c89b73898/lib/grape/attack/request.rb:25:in `instance_eval'
/home/vagrant/.rvm/gems/ruby-2.1.4/bundler/gems/grape-attack-a50c89b73898/lib/grape/attack/request.rb:25:in `client_identifier'
/home/vagrant/.rvm/gems/ruby-2.1.4/bundler/gems/grape-attack-a50c89b73898/lib/grape/attack/counter.rb:31:in `key'
/home/vagrant/.rvm/gems/ruby-2.1.4/bundler/gems/grape-attack-a50c89b73898/lib/grape/attack/counter.rb:14:in `value'
/home/vagrant/.rvm/gems/ruby-2.1.4/bundler/gems/grape-attack-a50c89b73898/lib/grape/attack/limiter.rb:39:in `allowed?'
/home/vagrant/.rvm/gems/ruby-2.1.4/bundler/gems/grape-attack-a50c89b73898/lib/grape/attack/limiter.rb:20:in `call!'
/home/vagrant/.rvm/gems/ruby-2.1.4/bundler/gems/grape-attack-a50c89b73898/lib/grape/attack/throttle.rb:9:in `before'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/grape-0.13.0/lib/grape/middleware/base.rb:23:in `call!'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/grape-0.13.0/lib/grape/middleware/base.rb:18:in `call'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/grape-0.13.0/lib/grape/middleware/error.rb:27:in `block in call!'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/grape-0.13.0/lib/grape/middleware/error.rb:26:in `catch'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/grape-0.13.0/lib/grape/middleware/error.rb:26:in `call!'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/grape-0.13.0/lib/grape/middleware/base.rb:18:in `call'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/rack-1.6.4/lib/rack/head.rb:13:in `call'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/rack-1.6.4/lib/rack/builder.rb:153:in `call'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/grape-0.13.0/lib/grape/endpoint.rb:202:in `call!'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/grape-0.13.0/lib/grape/endpoint.rb:190:in `call'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/rack-mount-0.8.3/lib/rack/mount/route_set.rb:152:in `block in call'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/rack-mount-0.8.3/lib/rack/mount/code_generation.rb:96:in `block in recognize'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/rack-mount-0.8.3/lib/rack/mount/code_generation.rb:68:in `optimized_each'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/rack-mount-0.8.3/lib/rack/mount/code_generation.rb:95:in `recognize'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/rack-mount-0.8.3/lib/rack/mount/route_set.rb:141:in `call'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/grape-0.13.0/lib/grape/api.rb:114:in `call'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/grape-0.13.0/lib/grape/api.rb:44:in `call!'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/grape-0.13.0/lib/grape/api.rb:39:in `call'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/actionpack-4.2.5/lib/action_dispatch/routing/mapper.rb:51:in `serve'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/actionpack-4.2.5/lib/action_dispatch/journey/router.rb:43:in `block in serve'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/actionpack-4.2.5/lib/action_dispatch/journey/router.rb:30:in `each'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/actionpack-4.2.5/lib/action_dispatch/journey/router.rb:30:in `serve'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/actionpack-4.2.5/lib/action_dispatch/routing/route_set.rb:817:in `call'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/rack-pjax-0.8.0/lib/rack/pjax.rb:12:in `call'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/omniauth-1.2.2/lib/omniauth/strategy.rb:186:in `call!'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/omniauth-1.2.2/lib/omniauth/strategy.rb:164:in `call'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/omniauth-1.2.2/lib/omniauth/strategy.rb:186:in `call!'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/omniauth-1.2.2/lib/omniauth/strategy.rb:164:in `call'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/omniauth-1.2.2/lib/omniauth/strategy.rb:186:in `call!'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/omniauth-1.2.2/lib/omniauth/strategy.rb:164:in `call'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/warden-1.2.3/lib/warden/manager.rb:35:in `block in call'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/warden-1.2.3/lib/warden/manager.rb:34:in `catch'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/warden-1.2.3/lib/warden/manager.rb:34:in `call'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/rack-1.6.4/lib/rack/etag.rb:24:in `call'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/rack-1.6.4/lib/rack/conditionalget.rb:25:in `call'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/rack-1.6.4/lib/rack/head.rb:13:in `call'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/remotipart-1.2.1/lib/remotipart/middleware.rb:27:in `call'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/actionpack-4.2.5/lib/action_dispatch/middleware/params_parser.rb:27:in `call'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/actionpack-4.2.5/lib/action_dispatch/middleware/flash.rb:260:in `call'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/rack-1.6.4/lib/rack/session/abstract/id.rb:225:in `context'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/rack-1.6.4/lib/rack/session/abstract/id.rb:220:in `call'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/actionpack-4.2.5/lib/action_dispatch/middleware/cookies.rb:560:in `call'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/activerecord-4.2.5/lib/active_record/query_cache.rb:36:in `call'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/activerecord-4.2.5/lib/active_record/connection_adapters/abstract/connection_pool.rb:653:in `call'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/actionpack-4.2.5/lib/action_dispatch/middleware/callbacks.rb:29:in `block in call'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/activesupport-4.2.5/lib/active_support/callbacks.rb:88:in `__run_callbacks__'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/activesupport-4.2.5/lib/active_support/callbacks.rb:778:in `_run_call_callbacks'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/activesupport-4.2.5/lib/active_support/callbacks.rb:81:in `run_callbacks'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/actionpack-4.2.5/lib/action_dispatch/middleware/callbacks.rb:27:in `call'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/rails-dev-tweaks-1.2.0/lib/rails_dev_tweaks/granular_autoload/middleware.rb:36:in `call'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/actionpack-4.2.5/lib/action_dispatch/middleware/remote_ip.rb:78:in `call'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/airbrake-4.3.4/lib/airbrake/rails/middleware.rb:13:in `call'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/actionpack-4.2.5/lib/action_dispatch/middleware/debug_exceptions.rb:17:in `call'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/actionpack-4.2.5/lib/action_dispatch/middleware/show_exceptions.rb:30:in `call'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/railties-4.2.5/lib/rails/rack/logger.rb:38:in `call_app'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/railties-4.2.5/lib/rails/rack/logger.rb:22:in `call'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/request_store-1.2.0/lib/request_store/middleware.rb:8:in `call'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/actionpack-4.2.5/lib/action_dispatch/middleware/request_id.rb:21:in `call'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/rack-1.6.4/lib/rack/methodoverride.rb:22:in `call'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/rack-1.6.4/lib/rack/runtime.rb:18:in `call'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/activesupport-4.2.5/lib/active_support/cache/strategy/local_cache_middleware.rb:28:in `call'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/rack-1.6.4/lib/rack/lock.rb:17:in `call'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/actionpack-4.2.5/lib/action_dispatch/middleware/static.rb:116:in `call'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/rack-1.6.4/lib/rack/sendfile.rb:113:in `call'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/rack-cors-0.4.0/lib/rack/cors.rb:80:in `call'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/rack-mini-profiler-0.9.7/lib/mini_profiler/profiler.rb:276:in `call'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/airbrake-4.3.4/lib/airbrake/user_informer.rb:16:in `_call'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/airbrake-4.3.4/lib/airbrake/user_informer.rb:12:in `call'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/railties-4.2.5/lib/rails/engine.rb:518:in `call'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/railties-4.2.5/lib/rails/application.rb:165:in `call'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/rack-1.6.4/lib/rack/content_length.rb:15:in `call'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/puma-2.15.3/lib/puma/server.rb:541:in `handle_request'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/puma-2.15.3/lib/puma/server.rb:388:in `process_client'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/puma-2.15.3/lib/puma/server.rb:270:in `block in run'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/puma-2.15.3/lib/puma/thread_pool.rb:106:in `call'
/home/vagrant/.rvm/gems/ruby-2.1.4/gems/puma-2.15.3/lib/puma/thread_pool.rb:106:in `block in spawn_thread'
```

**Comment by gottfrois** [comment]
can you show me how you use grape attack? the grape endpoint please

**Comment by confact** [comment]
@gottfrois here you go:

``` ruby
      resource :stations do
        desc "Return a list of stations"

        paginate per_page: 20, max_per_page: 30, offset: 0
        throttle max: Proc.new { ratelimit }, per: 1.month, identifier: Proc.new { apitoken.User.id }
        get do
          authenticate!
          logging("stations")
          stations = ::Station.where(accepted: 1)

          # Use `paginate` helper to execute kaminari methods
          # with arguments automatically passed from params
          present paginate(stations), with: V2::Entities::Station
        end
      end
```

And if you want to see ratelimit:

``` ruby
      # the limit for the users
      def ratelimit
        if apitoken.User.present? && apitoken.User.Plan.present?
          apitoken.User.Plan.limit
        else
          3000
        end
      end
```

**Comment by gottfrois** [comment]
`Proc.new { ratelimit }` this works for you? It should raise an exception since I don't handle procs here yet

**Comment by confact** [comment]
@gottfrois doesn't matter if i change `Proc.new { ratelimit }` to `3000` - It still give same error.

**Comment by confact** [comment]
@gottfrois i think i know the issue by testing `grape-throttle` and got the same issue. When use Proc.new i got no access to params so I couldn't call it. That means it fails for both the identifier and max in grape-attack. In grape-throttle i fix it by adding this:

``` ruby
      def request
        @request ||= ::Rack::Request.new(env)
      end

      def params
        @params ||= request.params.to_hash
      end
```

and add params as parameter:

``` ruby
limit = limit.call(params)
```

and the new Proc.new is like this:

``` ruby
Proc.new { |params| Apiv2helpers::ratelimit(params) }
```

You can see the code here: https://github.com/Dirble/grape-throttle/blob/master/lib/grape/middleware/throttle_middleware.rb

I would gladly use grape-attack because of your nice error handling and such. And i also think my hotfix can be better done.

**Comment by gottfrois** [comment]
Fixed in `v0.2.0`. The proc is now run in `Grape::Attack::Request` object context so that you have access to its methods. This `params` method in your usecase https://github.com/gottfrois/grape-attack/blob/da291d6029add159251ee14d9378b350e2e6c879/lib/grape/attack/request.rb#L24-L26

Thanks for this :)
